ops/terraform/services/pipeline/modules/bfd_pipeline_manager/README.md

```diff
@@ -0,0 +1,73 @@
+# `bfd_pipeline_manager` Sub-module
+
+This submodule defines a Lambda and its corresponding infrastructure. This Lambda,
+`bfd-${env}-pipeline-manager`, "manages" the BFD Pipeline Terraservice; it is invoked when new files
+are created in the `env`'s corresponding S3 pipeline/ETL bucket at specific paths:
+
+- `Synthetic/Incoming/`
+- `Synthetic/Done/`
+- `Incoming/`
+- `Done/`
+
+Objects uploaded to or moved to these paths in the ETL bucket will invoke this Lambda, which will
+then determine if the BFD CCW Pipeline should be started or not. This depends on a few factors:
+
+- If the file was uploaded to either of the `Incoming` folders, and there are no _ongoing_ or
+  _queued_ data loads _at all_ (including other groups), the `bfd-deploy-pipeline-terraservice`
+  Jenkins pipeline is started using the `bfd-job-broker` with the arguments set to _create_ the CCW
```

Comment:
Is it possible to have two instances start if two items are added in quick succession/at the same time? For instance, during synthea loads I might move 20+ folders of data into synthetic/incoming for all the future load data; is there a chance the lambda would trigger, queue the creation of the ccw pipeline instance, and before it gets created, see one of the next folders and start another?

It's pretty important we don't ever have two pipelines running for the same environment, so wanted to make sure this was explicitly not possible.

Reply:
Short answer: no, we cannot ever have two BFD Pipeline instances running at the same time in any given environment. This constraint is enforced by the `pipeline` Terraservice, as only one EC2 instance can exist in our state at any given time for an environment

**Long answer:**

It is _technically_ possible for the `bfd-pipeline-manager` _Lambda_ to run concurrently if S3 Event Notifications are received in quick succession and the locking SQS queue does not update fast enough. This would result in multiple `bfd-job-broker` executions, which in-turn would execute the `bfd-pipeline-terraservice` Jenkins job multiple times. Fortunately, the `bfd-pipeline-terraservice` Jenkins Pipeline uses Jenkins' built-in [Lockable Resources](https://www.jenkins.io/doc/pipeline/steps/lockable-resources/) such that, per-environment, only a single build of the `bfd-pipeline-terraservice` Jenkins Pipeline can be running at any time. Additionally, Terraform itself has a locking mechanism where only one operator/process can modify state at a time, so if the Jenkins lock failed this additional lock would stop concurrent modifications to the `pipeline` Terraservice state.

However, for the same reasons mentioned in the "short answer", only a single BFD CCW Pipeline instance can exist for a given environment at any time due to how its managed within the `pipeline` Terraservice. The only way this constraint could be broken is if the aforementioned locking mechanisms failed and the Terraform state for the `pipeline` Terraservice was somehow corrupted or emptied, thus allowing Terraform to create a new instance even if there was one already running. Of course, that would be a defect with Terraform, and not with this solution itself.